File: cpp/leetcode.525.cpp
#include <bits/stdc++.h>

using namespace std;

map<int, set<int>> mp;

class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        mp = map<int, set<int>>();
        int acc = 0;
        mp[0] = {-1};
        int mx = 0;

        for(int i = 0; i < n; i++){
            int delta = nums[i] == 0 ? -1 : 1;
            acc += delta;

            map<int, set<int>>::iterator iter = mp.find(acc);
            if (iter != mp.end()){
                mx = max(mx, i - *(iter->second.begin()) );
            }

            if (iter == mp.end()){
                mp[acc] = set<int>();
            }
            mp[acc].insert(i);
        }
        return mx;
    }
};

//int main()
//{
//    Solution sln;
//    vector<int> inp;
//    inp = {0, 1};
//    printf("%d\n", sln.findMaxLength(inp));
//    inp = {0, 1, 0};
//    printf("%d\n", sln.findMaxLength(inp));
//    inp = {0, 1, 1};
//    printf("%d\n", sln.findMaxLength(inp));
//    return 0;
//}

File: python/leetcode.525.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        delta = 0
        n = len(nums)
        st = {0:[-1]}

        mx = 0
        acc = 0
        for i in xrange(n):
            x = nums[i]
            delta = 0
            if x == 0:
                delta -= 1
            else:
                delta += 1

            acc += delta
            wanna_find = acc
            # print "acc", acc, "Wanna find", wanna_find, wanna_find in st
            if wanna_find in st:
                # print "==>", i, st[wanna_find]
                mx = max(mx, i - min(st[wanna_find]))

            # Update
            if not acc in st:
                st[acc] = []
            st[acc].append(i)

        return mx

# sln = Solution()
# print sln.findMaxLength([0,1]) # 2
# print sln.findMaxLength([0,1,0]) # 2
# print sln.findMaxLength([0,1,1]) # 2